## Python
* creator : [Guido van Rossum](https://en.wikipedia.org/wiki/Guido_van_Rossum), 1989
* [What is Pythonic?](https://blog.startifact.com/posts/older/what-is-pythonic.html)
* Interpreter 방식 : script language


## Python Tutorials

<ul>
  <li><a href="https://www.w3schools.com/python/" target="_blank">W3 School :  Python Tutorial</a> </li>
  <li><a href="https://perso.limsi.fr/pointal/_media/python:cours:mementopython3-english.pdf" target="_blank">Python 3 Cheat Sheet</a> </li>
    <li><a href="https://thepythonguru.com" target="_blank">
https://thepythonguru.com
</a> </li>
    <li><a href="https://docs.python.org/3/tutorial/" target="_blank">
https://docs.python.org/3/tutorial/
</a> </li>
</ul>

> Copyright (c) 2018 Hugon Kim(Ph.D) <br>
> [good47906@gmail.com](mailto:good47906@gmail.com) <br>
> [(주)에이옵트)](http://aopt.biz" target="_blank)<br>
> [mdhub.io](https://mdhub.io)

# Python package install
1. What is Anaconda and conda?
   * Conda is a package manager; Anaconda is a distribution. Although Conda is packaged with Anaconda, the two are distinct entities with distinct goals.
   * conda installs any package in conda environments.
   * \$conda install pkgname    
   * \$conda list    
2. PIP is a package manager for Python packages, or modules if you like. 
   * pip installs python packages in any environment.
   * pip now installs binary wheels if they are available, else it compiles everything from source.
   * [https://pypi.org/](https://pypi.org/)
   * \$pip install pkgname    
   * \$pip list ( or pip freeze)    
3. setup.py
   * \$git clone https://github.com/user/foo  
   * \$cd foo
   * \$python setup.py install

## Python package install via Jupyter Notebook


In [3]:
import numpy as np
np.__version__

'1.14.3'

In [5]:
import vega

ModuleNotFoundError: No module named 'vega'

In [ ]:
import sys
!conda install --yes --prefix {sys.prefix} numpy
!{sys.executable} -m pip install --upgrade pip
!{sys.executable} -m pip install xgboost
!{sys.executable} -m pip install shap

In [2]:
!type python

-a: not found
python is /usr/bin/python


In [3]:
!type conda

conda: not found


In [6]:
!python -m pip --version

/usr/bin/python: No module named pip


# virtualenv
* 서로 다양한 version의 package를 일괄적으로 관리하는데 필요함
* 예를 들어 
   * projectA에서 numpy=1.1, opencv=2.2 를 사용하고,
   * projectB에서 numpy=1.2, opencv=2.1 를 사용하는 경우
   * projectA, projectA를 위한 virtualenv를 만들어서 package를 설치하여 projectA, projectA를 각각 운영하면 됨

## windows :  run anaconda prompt
* (base) C:\Users\kim>
* (base) C:\Users\kim>conda create -n myenv   
* (base) C:\Users\kim>conda create -n myenv python=3.4 <===실행하지 말 것
* (base) C:\Users\kim>conda activate myenv
* (myenv) C:\Users\kim>
* (myenv) C:\Users\kim>conda install -n myenv keras
* (myenv) C:\Users\kim>conda install --yes --file requirements.txt
* (myenv) C:\Users\kim>jupyter notebook
* (myenv) C:\Users\kim>conda info --envs
* (myenv) C:\Users\kim>conda list
*  ...
* (myenv) C:\Users\kim>conda list -e > requirements.txt
* (myenv) C:\Users\kim> deactivate
* (base) C:\Users\kim>    

## Linux
* \$
* \$conda create -n myenv
* \$source activate myenv
* (myenv) \$
* (myenv) \$conda install -n myenv keras
* (myenv) \$conda install --yes --file requirements.txt
* (myenv) \$jupyter notebook
* (myenv) \$conda info --envs
* (myenv) \$conda list
*  ...
* (myenv) \$conda list -e > requirements.txt
* (myenv) \$source deactivate
* \$